LinuxAdapter.cs

```diff
@@ -0,0 +1,118 @@
+﻿using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using System;
+using System.DirectoryServices.Protocols;
+using System.Linq;
+using System.Net;
+using System.Security.Claims;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Authentication.Negotiate
+{
+    public class LinuxAdapter
+    {
+        private readonly string _distinguishedName;
+        private readonly LdapConnection _connection;
+
+        // example: domain = "DOMAIN.com"
+        public LinuxAdapter(string domain)
+        {
+            if (string.IsNullOrEmpty(domain))
+            {
+                throw new ArgumentNullException(nameof(domain));
+            }
+
+            _distinguishedName = domain.Split('.').Select(name => $"dc={name}").Aggregate((a, b) => $"{a},{b}");
+
+            LdapDirectoryIdentifier di = new LdapDirectoryIdentifier(server: domain, fullyQualifiedDnsHostName: true, connectionless: false);
+            _connection = new LdapConnection(di);
+            _connection.SessionOptions.ProtocolVersion = 3; //Setting LDAP Protocol to latest version
+            _connection.Timeout = TimeSpan.FromMinutes(1);
+            _connection.Bind(); // This line actually makes the connection.
+        }
+
+        public Task OnAuthenticated(AuthenticatedContext context, bool resolveNestedGroups = true)
+        {
+            var loggerFactory = context.HttpContext.RequestServices.GetRequiredService<ILoggerFactory>();
+            var logger = loggerFactory.CreateLogger("LinuxAdapter");
+
+            var user = context.Principal.Identity.Name;
+            var userAccountName = user.Substring(0, user.IndexOf('@'));
+
+            string filter = $"(&(objectClass=user)(sAMAccountName={userAccountName}))"; // This is using ldap search query language, it is looking on the server for someUser
+            SearchRequest searchRequest = new SearchRequest(_distinguishedName, filter, SearchScope.Subtree, null);
+            SearchResponse searchResponse = (SearchResponse)_connection.SendRequest(searchRequest);
```

Comment:
Might as well try the async version, that was fixed in preview6, right?
```suggestion
            var asyncResult = _connection.BeginSendRequest(searchRequest);
            SearchResponse searchResponse = (SearchResponse) await Task.Factory.FromAsync(asyncResult, _connection.EndSendRequest);
```

Reply:
Let me try this out first.